exonum-java-binding/core/src/test/java/com/exonum/binding/core/runtime/ServiceArtifactIdTest.java

```diff
@@ -41,25 +43,66 @@ void parseFromRoundtrip(String serviceId) {
   void valueOf() {
     int runtimeId = 1;
     String name = "test-service";
-    ServiceArtifactId id = ServiceArtifactId.valueOf(runtimeId, name);
+    String version = "0.1";
+    ServiceArtifactId id = ServiceArtifactId.valueOf(runtimeId, name, version);
 
     assertThat(id.getRuntimeId()).isEqualTo(runtimeId);
     assertThat(id.getName()).isEqualTo(name);
+    assertThat(id.getVersion()).isEqualTo(version);
   }
 
   @Test
   void newJavaId() {
     String name = "test-service";
-    ServiceArtifactId id = ServiceArtifactId.newJavaId(name);
+    String version = "test-version";
+    ServiceArtifactId id = ServiceArtifactId.newJavaId(name, version);
 
     assertThat(id.getRuntimeId()).isEqualTo(RuntimeId.JAVA.getId());
     assertThat(id.getName()).isEqualTo(name);
   }
 
+  @Test
+  void fromProto() {
+    ArtifactId artifactId = createArtifactIdMessage();
+
+    ServiceArtifactId serviceArtifactId = ServiceArtifactId.fromProto(artifactId);
+
+    assertThat(serviceArtifactId.getRuntimeId()).isEqualTo(artifactId.getRuntimeId());
+    assertThat(serviceArtifactId.getName()).isEqualTo(artifactId.getName());
+    assertThat(serviceArtifactId.getVersion()).isEqualTo(artifactId.getVersion());
+  }
+
   @Test
   void toStringTest() {
-    ServiceArtifactId id = ServiceArtifactId.valueOf(0, "full-name");
+    ServiceArtifactId id = ServiceArtifactId.valueOf(0, "full-name", "0.1");
 
-    assertThat(id.toString()).isEqualTo("0:full-name");
+    assertThat(id.toString()).isEqualTo("0:full-name:0.1");
   }
+
+  @ParameterizedTest
+  @ValueSource(strings = {
+      "",
+      "1:too-few:components:1.0",
+      "1:com.acme:foo-service:0.1.0:extra-component",
+      " : : ",
+      "1:com acme:foo:1.0",
+      "1 :com.acme:foo:1.0",
+      "1:com.acme:foo service:1.0",
+      "1:com.acme:foo-service:1 0",
+      "1:com.acme:foo-service: 1.0",
+      "1:com.acme:foo-service:1.0 ",
```

Comment:
> The sources need to be updated to reflect the new string representation.
E.g., "too-few:components:1.0" — is fine in terms of components; the problem here is non-integral runtime id.
Or, several "com.acme:foo service:1.0" with spaces do not necessarily trigger that as they, again,
have an invalid runtime id.

It still doesn't make sense — each element in the test data must have a single problem, and otherwise be valid. FTFY:

```java
      "",
      /* Too few components */
      "too-few-components",
      "1:too-few-components",
      /* Extra component */
      "1:foo-service:0.1.0:extra-component",
      /* All blanks */
      " : : ",
      /* Non-integral runtime id */
      "a:com.acme/foo:1.0",
      /* Spaces in runtime id */
      "1 :com.acme/foo:1.0",
      /* Spaces in name */
      "1:com.acme foo:1.0",
      "1:com.acme/fo o:1.0",
      /* Spaces in version */
      "1:com.acme:foo: 1.0",
      "1:com.acme/foo:1.0 ",
      "1:com.acme:foo:1 0",
```